pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,242 @@
+				service := cur.(*kapi.Service)
+				glog.V(4).Infof("Updating service %s", service.Name)
+				// Resync on service object relist.
+				sc.enqueueService(cur)
+			},
+			// This will enter the sync loop and no-op, because the service has been deleted from the store.
+			DeleteFunc: func(obj interface{}) {
+				service := obj.(*kapi.Service)
+				glog.V(4).Infof("Deleting service %s", service.Name)
+				sc.enqueueService(obj)
+			},
+		},
+	)
+
+	return sc
+}
+
+// Run begins watching and syncing.
+func (sc *ServiceServingCertController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.Infof("Shutting down service signing cert controller")
+	sc.queue.ShutDown()
+}
+
+// obj could be an *kapi.Service, or a DeletionFinalStateUnknown marker item.
+func (sc *ServiceServingCertController) enqueueService(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+
+	// TODO: Handle overlapping services better. Either disallow them at admission time or
+	// deterministically avoid syncing services that fight over ReplicaSet's. Currently, we
+	// only ensure that the same service is synced for a given ReplicaSet. When we
+	// periodically relist all services there will still be some ReplicaSet instability. One
+	//  way to handle this is by querying the store for all services that this service
+	// overlaps, as well as all services that overlap this services, and sorting them.
+	sc.queue.Add(key)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceServingCertController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncService(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing service: %v", err)
+			}
+		}()
+	}
+}
+
+// syncService will sync the service with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (sc *ServiceServingCertController) syncService(key string) error {
+	obj, exists, err := sc.serviceCache.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		sc.queue.Add(key)
+		return err
+	}
+	if !exists {
+		glog.Infof("Service has been deleted %v", key)
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	t := *obj.(*kapi.Service)
+	service := &t
+
+	if !sc.requiresCertGeneration(service) {
+		return nil
+	}
+
+	dnsName := service.Name + "." + service.Namespace + ".svc"
+	fqDNSName := dnsName + "." + sc.dnsSuffix
+	servingCert, err := sc.ca.MakeServerCertWithoutWriting(sets.NewString(dnsName, fqDNSName))
```

Comment:
Lack of trust hurts!  Service ips are reused across namespaces so it's not safe to sign them.